[docker-compose-recipes] Add docker compose recipes

nellicus opened the pull request:
This PR adds the following docker compose examples.

A list of ClickHouse docker compose recipes

- ClickHouse and Grafana
- ClickHouse and Minio S3
- Clickhouse and LDAP (OpenLDAP) - WIP
- Clickhouse Cluster: 2 CH nodes - 3 ClickHouse Keeper (1 Shard 2 Replicas)
- Clickhouse Cluster: 2 CH nodes - 3 ClickHouse Keeper (2 Shards 1 Replica)
- Clickhouse Cluster: 4 CH nodes - 3 ClickHouse Keeper (2 Shards 2 Replicas)
- Clickhouse Cluster: 2 CH nodes - 3 ClickHouse Keeper (1 Shard 2 Replicas) - CH Proxy LB
- Clickhouse Cluster: 2 CH nodes - 3 ClickHouse Keeper (2 Shards 1 Replica) - CH Proxy LB
- Clickhouse Cluster: 4 CH nodes - 3 ClickHouse Keeper (2 Shards 2 Replicas) - CH Proxy LB

README file are added and updated and a minor repo folder reorg to allow for this new content.

DanRoscigno commented:
One concern: The keeper containers are running Keeper embedded in the ClickHouse server process rather than standalone Keeper.  We recommend running standalone Keepers, and the config files would then be put in `/etc/clickhouse-keeper/` instead. 

@felixoid do you think that we should have a `latest` tag for the `clickhouse-keeper` Docker image?  There is a `latest-alpine`, but not plain `latest`.

Felixoid commented:
It's possible to add, although the hack looks a bit dirty to me. It's on the way

DanRoscigno commented on docker-compose-recipes/cluster_1S_2R/docker-compose.yaml:
```diff
@@ -0,0 +1,97 @@
+      - ${PWD}/fs/volumes/clickhouse-keeper-03/var/log/clickhouse-server:/var/log/clickhouse-server
+    ports:
+        - "127.0.0.1:8129:8123"
+        - "127.0.0.1:9006:9000"
```
I think that on the keeper servers exposing 9181 would be good to allow running `echo mntr | nc localhost 9183` from the local machine.  9000 and 8123 probably do not need to be exposed since clickhouse server on the Keeper nodes is not being used.  (I will probably submit a PR to switch to using the clickhouse-keeper Docker image for the three Keeper nodes, but when I am trying this it is running the process as user root instead of user clickhouse, and I have not figured out why yet)

DanRoscigno commented on docker-compose-recipes/cluster_1S_2R/fs/volumes/clickhouse-keeper-01/etc/clickhouse-server/users.d/users.xml:
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0"?>
```
I believe that on standalone keeper servers this file should not exist.  I looked in the package files for the clickhouse-keeper distribution files and there is only `keeper_configxml`

nellicus commented:
one undesirable effect is taht the terminal where the recipe is spun up is now flooded with keeper junk logs

```
house-keeper-03  | (update) new target priority: 1
clickhouse-keeper-02  | EMPTY PAYLOAD
clickhouse-keeper-03  | [LOG OK] req log idx: 9, req log term: 1, my last log idx: 9, my log (9) term: 1
clickhouse-keeper-02  | send req 2 -> 1, type append_entries_request
clickhouse-keeper-03  | local log idx 9, target_commit_idx 9, quick_commit_index_ 9, state_->get_commit_idx() 9
clickhouse-keeper-02  | sent
clickhouse-keeper-03  | cancel existing timer
clickhouse-keeper-03  | re-schedule election timer
clickhouse-keeper-03  | batch size hint: 0 bytes
clickhouse-keeper-03  | Response back a append_entries_response message to 2 with Accepted=1, Term=1, NextIndex=10
clickhouse-keeper-02  | resp of req 2 -> 3, type append_entries_request, OK
clickhouse-keeper-02  | Receive a append_entries_response message from peer 3 with Result=1, Term=1, NextIndex=10
clickhouse-keeper-02  | src: 3, dst: 2, resp->get_term(): 1
clickhouse-keeper-02  | handle append entries resp (from 3), resp.get_next_idx(): 10
clickhouse-keeper-02  | peer 3 batch size hint: 0 bytes
clickhouse-keeper-02  | peer 3, prev matched idx: 9, new matched idx: 9
clickhouse-keeper-02  | last durable index 9, precommit index 9
clickhouse-keeper-02  | quorum idx 1, 9 9 9
clickhouse-keeper-02  | local log idx 9, target_commit_idx 9, quick_commit_index_ 9, state_->get_commit_idx() 9
clickhouse-keeper-02  | resp of req 2 -> 1, type append_entries_request, OK
clickhouse-keeper-02  | Receive a append_entries_response message from peer 1 with Result=1, Term=1, NextIndex=10
clickhouse-keeper-02  | src: 1, dst: 2, resp->get_term(): 1
clickhouse-keeper-02  | handle append entries resp (from 1), resp.get_next_idx(): 10
clickhouse-keeper-02  | peer 1 batch size hint: 0 bytes
clickhouse-keeper-02  | peer 1, prev matched idx: 9, new matched idx: 9
clickhouse-keeper-02  | last durable index 9, precommit index 9
clickhouse-keeper-02  | quorum idx 1, 9 9 9
clickhouse-keeper-02  | local log idx 9, target_commit_idx 9, quick_commit_index_ 9, state_->get_commit_idx() 9
clickhouse-keeper-01  | Receive a append_entries_request message fr
```

Felixoid commented:
The `clickhouse/clickhouse-keeper:latest` is here https://hub.docker.com/r/clickhouse/clickhouse-keeper/tags?page=1&name=latest

nellicus commented:
@DanRoscigno  as of now with my latest commit we get:

cluster_1S_2R docker compose up

```
Attaching to ch-proxy, clickhouse-01, clickhouse-02, clickhouse-keeper-01, clickhouse-keeper-02, clickhouse-keeper-03
clickhouse-keeper-01  | Processing configuration file '/etc/clickhouse-keeper/keeper_config.xml'.
clickhouse-keeper-02  | Processing configuration file '/etc/clickhouse-keeper/keeper_config.xml'.
clickhouse-keeper-03  | Processing configuration file '/etc/clickhouse-keeper/keeper_config.xml'.
clickhouse-keeper-01  | Starting ClickHouse Keeper 23.3.1.2823(revision : 54472, git hash: 46e85357ce2da2a99f56ee83a079e892d7ec3726, build id: 27B759B5A042E67EE0BC17B859AFF90A614EB343), PID 1
clickhouse-keeper-01  | starting up
clickhouse-keeper-01  | OS Name = Linux, OS Version = 5.15.49-linuxkit, OS Architecture = aarch64
clickhouse-keeper-02  | Starting ClickHouse Keeper 23.3.1.2823(revision : 54472, git hash: 46e85357ce2da2a99f56ee83a079e892d7ec3726, build id: 27B759B5A042E67EE0BC17B859AFF90A614EB343), PID 1
clickhouse-keeper-02  | starting up
clickhouse-keeper-02  | OS Name = Linux, OS Version = 5.15.49-linuxkit, OS Architecture = aarch64
clickhouse-keeper-03  | Starting ClickHouse Keeper 23.3.1.2823(revision : 54472, git hash: 46e85357ce2da2a99f56ee83a079e892d7ec3726, build id: 27B759B5A042E67EE0BC17B859AFF90A614EB343), PID 1
clickhouse-keeper-03  | starting up
clickhouse-keeper-03  | OS Name = Linux, OS Version = 5.15.49-linuxkit, OS Architecture = aarch64
clickhouse-keeper-01  | force_sync enabled
clickhouse-keeper-02  | force_sync enabled
clickhouse-keeper-01  | Preprocessing 125 log entries
clickhouse-keeper-01  | Preprocessing done
clickhouse-keeper-01  | Will use config from log store with log index 1838
clickhouse-keeper-01  | Raft ASIO listener initiated on :::9234, unsecured
clickhouse-keeper-03  | force_sync enabled
clickhouse-keeper-01  | Read state from /var/lib/clickhouse/coordination/state
clickhouse-keeper-01  | parameters: election timeout range 1000 - 2000, heartbeat 500, leadership expiry 10000, max batch 100, backoff 50, snapshot distance 100000, enable randomized snapshot creation NO, log sync stop gap 99999, reserved logs 100000, client timeout 10000, auto forwarding on, API call type async, custom commit quorum size 0, custom election quorum size 0, snapshot receiver included, leadership transfer wait time 0, grace period of lagging state machine 0, snapshot IO: blocking, parallel log appending: on
clickhouse-keeper-01  | new election timeout range: 1000 - 2000
clickhouse-keeper-01  |    === INIT RAFT SERVER ===
clickhouse-keeper-01  | commit index 2122
clickhouse-keeper-01  | term 14
clickhouse-keeper-01  | election timer allowed
clickhouse-keeper-01  | log store start 1, end 2247
clickhouse-keeper-01  | config log idx 2123, prev log idx 1838
clickhouse-keeper-01  | detect a configuration change that is not committed yet at index 2123
clickhouse-keeper-01  | peer 1: DC ID 0, clickhouse-keeper-01:9234, voting member, 1
clickhouse-keeper-01  | peer 2: DC ID 0, clickhouse-keeper-02:9234, voting member, 1
clickhouse-keeper-01  | peer 3: DC ID 0, clickhouse-keeper-03:9234, voting member, 1
clickhouse-keeper-01  | my id: 1, voting_member
clickhouse-keeper-01  | num peers: 2
clickhouse-keeper-01  | global manager does not exist. will use local thread for commit and append
clickhouse-keeper-01  | wait for HB, for 50 + [1000, 2000] ms
clickhouse-keeper-01  | bg append_entries thread initiated
clickhouse-keeper-02  | Preprocessing 125 log entries
clickhouse-keeper-02  | Preprocessing done
clickhouse-keeper-02  | Will use config from log store with log index 1838
clickhouse-keeper-02  | Raft ASIO listener initiated on :::9234, unsecured
clickhouse-keeper-03  | Preprocessing 125 log entries
clickhouse-keeper-03  | Preprocessing done
clickhouse-keeper-03  | Will use config from log store with log index 1838
clickhouse-keeper-02  | Read state from /var/lib/clickhouse/coordination/state
clickhouse-keeper-03  | Raft ASIO listener initiated on :::9234, unsecured
clickhouse-keeper-02  | parameters: election timeout range 1000 - 2000, heartbeat 500, leadership expiry 10000, max batch 100, backoff 50, snapshot distance 100000, enable randomized snapshot creation NO, log sync stop gap 99999, reserved logs 100000, client timeout 10000, auto forwarding on, API call type async, custom commit quorum size 0, custom election quorum size 0, snapshot receiver included, leadership transfer wait time 0, grace period of lagging state machine 0, snapshot IO: blocking, parallel log appending: on
clickhouse-keeper-02  | new election timeout range: 1000 - 2000
clickhouse-keeper-02  |    === INIT RAFT SERVER ===
clickhouse-keeper-02  | commit index 2122
clickhouse-keeper-02  | term 14
clickhouse-keeper-02  | election timer allowed
clickhouse-keeper-02  | log store start 1, end 2247
clickhouse-keeper-02  | config log idx 2123, prev log idx 1838
clickhouse-keeper-02  | detect a configuration change that is not committed yet at index 2123
clickhouse-keeper-02  | peer 1: DC ID 0, clickhouse-keeper-01:9234, voting member, 1
clickhouse-keeper-02  | peer 2: DC ID 0, clickhouse-keeper-02:9234, voting member, 1
clickhouse-keeper-02  | peer 3: DC ID 0, clickhouse-keeper-03:9234, voting member, 1
clickhouse-keeper-02  | my id: 2, voting_member
clickhouse-keeper-02  | num peers: 2
clickhouse-keeper-02  | global manager does not exist. will use local thread for commit and append
clickhouse-keeper-02  | wait for HB, for 50 + [1000, 2000] ms
clickhouse-keeper-02  | bg append_entries thread initiated
clickhouse-keeper-03  | Read state from /var/lib/clickhouse/coordination/state
clickhouse-keeper-03  | parameters: election timeout range 1000 - 2000, heartbeat 500, leadership expiry 10000, max batch 100, backoff 50, snapshot distance 100000, enable randomized snapshot creation NO, log sync stop gap 99999, reserved logs 100000, client timeout 10000, auto forwarding on, API call type async, custom commit quorum size 0, custom election quorum size 0, snapshot receiver included, leadership transfer wait time 0, grace period of lagging state machine 0, snapshot IO: blocking, parallel log appending: on
clickhouse-keeper-03  | new election timeout range: 1000 - 2000
clickhouse-keeper-03  |    === INIT RAFT SERVER ===
clickhouse-keeper-03  | commit index 2122
clickhouse-keeper-03  | term 14
clickhouse-keeper-03  | election timer allowed
clickhouse-keeper-03  | log store start 1, end 2247
clickhouse-keeper-03  | config log idx 2123, prev log idx 1838
clickhouse-keeper-03  | detect a configuration change that is not committed yet at index 2123
clickhouse-keeper-03  | peer 1: DC ID 0, clickhouse-keeper-01:9234, voting member, 1
clickhouse-keeper-03  | peer 2: DC ID 0, clickhouse-keeper-02:9234, voting member, 1
clickhouse-keeper-03  | peer 3: DC ID 0, clickhouse-keeper-03:9234, voting member, 1
clickhouse-keeper-03  | my id: 3, voting_member
clickhouse-keeper-03  | num peers: 2
clickhouse-keeper-03  | global manager does not exist. will use local thread for commit and append
clickhouse-keeper-03  | wait for HB, for 50 + [1000, 2000] ms
clickhouse-keeper-03  | bg append_entries thread initiated
clickhouse-keeper-01  | Starting Keeper asynchronously, server will accept connections to Keeper when it will be ready
clickhouse-keeper-01  | Server still not initialized, will not apply configuration until initialization finished
clickhouse-keeper-01  | Listening for Keeper (tcp): 0.0.0.0:9181
clickhouse-keeper-01  | Ready for connections.
clickhouse-keeper-02  | Starting Keeper asynchronously, server will accept connections to Keeper when it will be ready
clickhouse-keeper-02  | Listening for Keeper (tcp): 0.0.0.0:9181
clickhouse-keeper-02  | Server still not initialized, will not apply configuration until initialization finished
clickhouse-keeper-03  | Starting Keeper asynchronously, server will accept connections to Keeper when it will be ready
clickhouse-keeper-03  | Server still not initialized, will not apply configuration until initialization finished
clickhouse-keeper-03  | Listening for Keeper (tcp): 0.0.0.0:9181
clickhouse-keeper-02  | Ready for connections.
clickhouse-keeper-03  | Ready for connections.
ch-proxy              | INFO: 2023/04/03 12:59:47 main.go:47: chproxy ver. 1.22.0, rev. 5c1e8e711fba637ebc725bd1195acf52cd93a3f1, built at 2023-03-01T09:04:29Z
ch-proxy              | INFO: 2023/04/03 12:59:47 main.go:48: Loading config: /opt/config.yml
ch-proxy              | ERROR: 2023/04/03 12:59:47 scope.go:734: error while health-checking "172.29.5.2:8123" host: cannot send request in 25.295292ms: Get "http://172.29.5.2:8123/ping": dial tcp 172.29.5.2:8123: connect: connection refused
ch-proxy              | ERROR: 2023/04/03 12:59:47 scope.go:734: error while health-checking "172.29.5.1:8123" host: cannot send request in 25.286042ms: Get "http://172.29.5.1:8123/ping": dial tcp 172.29.5.1:8123: connect: connection refused
ch-proxy              | INFO: 2023/04/03 12:59:47 main.go:305: Loaded config:
ch-proxy              | server:
ch-proxy              |   http:
ch-proxy              |     listen_addr: :9090
ch-proxy              |     allowed_networks:
ch-proxy              |     - 127.0.0.0/24
ch-proxy              |     - 172.29.0.0/16
ch-proxy              |     forceautocerthandler: false
ch-proxy              |     read_timeout: 1m
ch-proxy              |     write_timeout: 3m
ch-proxy              |     idle_timeout: 10m
ch-proxy              | clusters:
ch-proxy              | - name: default
ch-proxy              |   scheme: http
ch-proxy              |   nodes:
ch-proxy              |   - 172.29.5.1:8123
ch-proxy              |   - 172.29.5.2:8123
ch-proxy              |   users:
ch-proxy              |   - name: default
ch-proxy              |     password: XXX
ch-proxy              |   heartbeat:
ch-proxy              |     interval: 5s
ch-proxy              |     timeout: 3s
ch-proxy              |     request: /ping
ch-proxy              |     response: |
ch-proxy              |       Ok.
ch-proxy              | users:
ch-proxy              | - name: default
ch-proxy              |   password: XXX
ch-proxy              |   to_cluster: default
ch-proxy              |   to_user: default
ch-proxy              |   max_execution_time: 2m
ch-proxy              | INFO: 2023/04/03 12:59:47 main.go:59: Loading config "/opt/config.yml": successful
ch-proxy              | INFO: 2023/04/03 12:59:47 main.go:165: Serving http on ":9090"
clickhouse-01         | Processing configuration file '/etc/clickhouse-server/config.xml'.
clickhouse-01         | Merging configuration file '/etc/clickhouse-server/config.d/config.xml'.
clickhouse-01         | Logging debug to /var/log/clickhouse-server/clickhouse-server.log
clickhouse-01         | Logging errors to /var/log/clickhouse-server/clickhouse-server.err.log
clickhouse-01         | Processing configuration file '/etc/clickhouse-server/config.xml'.
clickhouse-01         | Merging configuration file '/etc/clickhouse-server/config.d/config.xml'.
clickhouse-01         | Saved preprocessed configuration to '/var/lib/clickhouse/preprocessed_configs/config.xml'.
clickhouse-01         | Processing configuration file '/etc/clickhouse-server/users.xml'.
clickhouse-01         | Merging configuration file '/etc/clickhouse-server/users.d/users.xml'.
clickhouse-01         | Saved preprocessed configuration to '/var/lib/clickhouse/preprocessed_configs/users.xml'.
clickhouse-02         | Processing configuration file '/etc/clickhouse-server/config.xml'.
clickhouse-02         | Merging configuration file '/etc/clickhouse-server/config.d/config.xml'.
clickhouse-02         | Logging debug to /var/log/clickhouse-server/clickhouse-server.log
clickhouse-02         | Logging errors to /var/log/clickhouse-server/clickhouse-server.err.log
clickhouse-02         | Processing configuration file '/etc/clickhouse-server/config.xml'.
clickhouse-02         | Merging configuration file '/etc/clickhouse-server/config.d/config.xml'.
clickhouse-02         | Saved preprocessed configuration to '/var/lib/clickhouse/preprocessed_configs/config.xml'.
clickhouse-02         | Processing configuration file '/etc/clickhouse-server/users.xml'.
clickhouse-02         | Merging configuration file '/etc/clickhouse-server/users.d/users.xml'.
clickhouse-02         | Saved preprocessed configuration to '/var/lib/clickhouse/preprocessed_configs/users.xml'.
clickhouse-keeper-01  | Election timeout, initiate leader election
clickhouse-keeper-01  | [PRIORITY] decay, target 1 -> 1, mine 1
clickhouse-keeper-01  | [ELECTION TIMEOUT] current role: follower, log last term 14, state term 14, target p 1, my p 1, hb dead, pre-vote NOT done
clickhouse-keeper-01  | pre-vote term (0) is different, reset it to 14
clickhouse-keeper-01  | [PRE-VOTE INIT] my id 1, my role candidate, term 14, log idx 2247, log term 14, priority (target 1 / mine 1)
clickhouse-keeper-01  | 0xffff00c40e18 connected to clickhouse-keeper-03:9234 (as a client)
clickhouse-keeper-01  | 0xffff00c40718 connected to clickhouse-keeper-02:9234 (as a client)
clickhouse-keeper-01  | [PRE-VOTE RESP] peer 3 (O), term 14, resp term 14, my role candidate, dead 2, live 0, num voting members 3, quorum 2
clickhouse-keeper-01  | [PRE-VOTE DONE] SUCCESS, term 14
clickhouse-keeper-01  | [PRE-VOTE DONE] initiate actual vote
clickhouse-keeper-03  | receive a incoming rpc connection
clickhouse-keeper-03  | session 1 got connection from ::ffff:172.29.5.5:39160 (as a server)
clickhouse-keeper-02  | receive a incoming rpc connection
clickhouse-keeper-02  | session 1 got connection from ::ffff:172.29.5.5:57474 (as a server)
clickhouse-keeper-02  | [PRE-VOTE REQ] my role follower, from peer 1, log term: req 14 / mine 14
clickhouse-keeper-02  | last idx: req 2247 / mine 2247, term: req 14 / mine 14
clickhouse-keeper-02  | HB dead
clickhouse-keeper-02  | pre-vote decision: O (grant)
clickhouse-keeper-03  | [PRE-VOTE REQ] my role follower, from peer 1, log term: req 14 / mine 14
clickhouse-keeper-03  | last idx: req 2247 / mine 2247, term: req 14 / mine 14
clickhouse-keeper-03  | HB dead
clickhouse-keeper-03  | pre-vote decision: O (grant)
clickhouse-keeper-01  | [VOTE INIT] my id 1, my role candidate, term 15, log idx 2247, log term 14, priority (target 1 / mine 1)
clickhouse-keeper-01  | [PRE-VOTE RESP] peer 2 (O), term 14, resp term 14, my role candidate, dead 3, live 0, num voting members 3, quorum 2
clickhouse-keeper-01  | [PRE-VOTE DONE] SUCCESS, term 14
clickhouse-keeper-01  | [PRE-VOTE DONE] actual vote is already initiated, do nothing
clickhouse-keeper-03  | [BECOME FOLLOWER] term 15
clickhouse-keeper-03  | [PRIORITY] decay, target 1 -> 1, mine 1
clickhouse-keeper-03  | [VOTE REQ] my role follower, from peer 1, log term: req 14 / mine 14
clickhouse-keeper-03  | last idx: req 2247 / mine 2247, term: req 15 / mine 15
clickhouse-keeper-03  | priority: target 1 / mine 1, voted_for -1
clickhouse-keeper-03  | decision: O (grant), voted_for 1, term 15
clickhouse-keeper-02  | [BECOME FOLLOWER] term 15
clickhouse-keeper-02  | [PRIORITY] decay, target 1 -> 1, mine 1
clickhouse-keeper-02  | [VOTE REQ] my role follower, from peer 1, log term: req 14 / mine 14
clickhouse-keeper-02  | last idx: req 2247 / mine 2247, term: req 15 / mine 15
clickhouse-keeper-02  | priority: target 1 / mine 1, voted_for -1
clickhouse-keeper-02  | decision: O (grant), voted_for 1, term 15
clickhouse-keeper-01  | [VOTE RESP] peer 2 (O), resp term 15, my role candidate, granted 2, responded 2, num voting members 3, quorum 2
clickhouse-keeper-01  | Server is elected as leader for term 15
clickhouse-keeper-01  | number of pending commit elements: 0
clickhouse-keeper-01  | state machine commit index 2122, precommit index 2247, last log index 2247
clickhouse-keeper-01  | found uncommitted config at 2123, size 145
clickhouse-keeper-01  | [BECOME LEADER] appended new config at 2248
clickhouse-keeper-01  | Could not preallocate space on disk using fallocate. Error: errno: 95, strerror: Operation not supported, errno: 95
clickhouse-keeper-01  |   === LEADER (term 15) ===
clickhouse-keeper-01  | Election completed, will ignore the voting result from this server
clickhouse-keeper-03  | Could not preallocate space on disk using fallocate. Error: errno: 95, strerror: Operation not supported, errno: 95
clickhouse-keeper-02  | Could not preallocate space on disk using fallocate. Error: errno: 95, strerror: Operation not supported, errno: 95
clickhouse-keeper-02  | receive a config change from leader at 2248
clickhouse-keeper-03  | receive a config change from leader at 2248
clickhouse-keeper-02  | config at index 2123 is committed, prev config log idx 2123
clickhouse-keeper-02  | skipped config 2123, latest config 2123
clickhouse-keeper-01  | config at index 2123 is committed, prev config log idx 2123
clickhouse-keeper-01  | skipped config 2123, latest config 2123
clickhouse-keeper-01  | config at index 2248 is committed, prev config log idx 2123
clickhouse-keeper-01  | new config log idx 2248, prev log idx 2123, cur config log idx 2123, prev log idx 1838
clickhouse-keeper-01  | new configuration: log idx 2248, prev log idx 2123
clickhouse-keeper-01  | peer 1, DC ID 0, clickhouse-keeper-01:9234, voting member, 1
clickhouse-keeper-01  | peer 2, DC ID 0, clickhouse-keeper-02:9234, voting member, 1
clickhouse-keeper-01  | peer 3, DC ID 0, clickhouse-keeper-03:9234, voting member, 1
clickhouse-keeper-01  | my id: 1, leader: 1, term: 15
clickhouse-keeper-03  | config at index 2123 is committed, prev config log idx 2123
clickhouse-keeper-03  | skipped config 2123, latest config 2123
clickhouse-keeper-02  | config at index 2248 is committed, prev config log idx 2123
clickhouse-keeper-02  | new config log idx 2248, prev log idx 2123, cur config log idx 2123, prev log idx 1838
clickhouse-keeper-02  | new configuration: log idx 2248, prev log idx 2123
clickhouse-keeper-02  | peer 1, DC ID 0, clickhouse-keeper-01:9234, voting member, 1
clickhouse-keeper-02  | peer 2, DC ID 0, clickhouse-keeper-02:9234, voting member, 1
clickhouse-keeper-02  | peer 3, DC ID 0, clickhouse-keeper-03:9234, voting member, 1
clickhouse-keeper-02  | my id: 2, leader: 1, term: 15
clickhouse-keeper-03  | config at index 2248 is committed, prev config log idx 2123
clickhouse-keeper-03  | new config log idx 2248, prev log idx 2123, cur config log idx 2123, prev log idx 1838
clickhouse-keeper-03  | new configuration: log idx 2248, prev log idx 2123
clickhouse-keeper-03  | peer 1, DC ID 0, clickhouse-keeper-01:9234, voting member, 1
clickhouse-keeper-03  | peer 2, DC ID 0, clickhouse-keeper-02:9234, voting member, 1
clickhouse-keeper-03  | peer 3, DC ID 0, clickhouse-keeper-03:9234, voting member, 1
clickhouse-keeper-03  | my id: 3, leader: 1, term: 15
clickhouse-keeper-02  | Requesting session ID for the new client
clickhouse-keeper-01  | receive a incoming rpc connection
clickhouse-keeper-02  | 0xffff00c6bc18 connected to clickhouse-keeper-01:9234 (as a client)
clickhouse-keeper-01  | session 1 got connection from ::ffff:172.29.5.6:47932 (as a server)
clickhouse-keeper-02  | Received session ID 27
clickhouse-keeper-03  | Requesting session ID for the new client
clickhouse-keeper-01  | receive a incoming rpc connection
clickhouse-keeper-01  | session 2 got connection from ::ffff:172.29.5.7:54152 (as a server)
clickhouse-keeper-03  | 0xfffdc0380018 connected to clickhouse-keeper-01:9234 (as a client)
clickhouse-keeper-03  | Received session ID 28
clickhouse-keeper-01  | Found dead session 26, will try to close it
clickhouse-keeper-01  | Dead session close request pushed
clickhouse-keeper-01  | Found dead session 25, will try to close it
clickhouse-keeper-01  | Dead session close request pushed
```

what would your proposal achieve instead?

DanRoscigno commented:
Fixed perms:
```
❯ sudo chown -R 101:101 fs
❯ ls -l fs/volumes/clickhouse-keeper-01/var/lib/
❯ docker compose up
[+] Running 6/4
 ⠿ Network network_cluster_1S_2R   Created                                                                   0.1s
 ⠿ Container clickhouse-keeper-02  Created                                                                   0.0s
 ⠿ Container clickhouse-keeper-01  Created                                                                   0.0s
 ⠿ Container clickhouse-keeper-03  Created                                                                   0.0s
 ⠿ Container clickhouse-02         Created                                                                   0.0s
 ⠿ Container clickhouse-01         Created                                                                   0.0s
Attaching to clickhouse-01, clickhouse-02, clickhouse-keeper-01, clickhouse-keeper-02, clickhouse-keeper-03
clickhouse-keeper-01  | Processing configuration file '/etc/clickhouse-keeper/keeper_config.xml'.
clickhouse-keeper-01  | Logging trace to /var/log/clickhouse-keeper/clickhouse-keeper.log
clickhouse-keeper-01  | Logging errors to /var/log/clickhouse-keeper/clickhouse-keeper.err.log
clickhouse-keeper-03  | Processing configuration file '/etc/clickhouse-keeper/keeper_config.xml'.
clickhouse-keeper-03  | Logging trace to /var/log/clickhouse-keeper/clickhouse-keeper.log
clickhouse-keeper-03  | Logging errors to /var/log/clickhouse-keeper/clickhouse-keeper.err.log
clickhouse-keeper-02  | Processing configuration file '/etc/clickhouse-keeper/keeper_config.xml'.
clickhouse-keeper-02  | Logging trace to /var/log/clickhouse-keeper/clickhouse-keeper.log
clickhouse-keeper-02  | Logging errors to /var/log/clickhouse-keeper/clickhouse-keeper.err.log
clickhouse-keeper-01  | Processing configuration file '/etc/clickhouse-keeper/keeper_config.xml'.
clickhouse-keeper-01  | Saved preprocessed configuration to '/etc/clickhouse-keeper/keeper_config-preprocessed.xml'.
clickhouse-keeper-03  | Processing configuration file '/etc/clickhouse-keeper/keeper_config.xml'.
clickhouse-keeper-03  | Saved preprocessed configuration to '/etc/clickhouse-keeper/keeper_config-preprocessed.xml'.
clickhouse-keeper-02  | Processing configuration file '/etc/clickhouse-keeper/keeper_config.xml'.
clickhouse-keeper-02  | Saved preprocessed configuration to '/etc/clickhouse-keeper/keeper_config-preprocessed.xml'.
clickhouse-02         | Processing configuration file '/etc/clickhouse-server/config.xml'.
clickhouse-02         | Merging configuration file '/etc/clickhouse-server/config.d/config.xml'.
clickhouse-02         | Logging debug to /var/log/clickhouse-server/clickhouse-server.log
clickhouse-02         | Logging errors to /var/log/clickhouse-server/clickhouse-server.err.log
clickhouse-01         | Processing configuration file '/etc/clickhouse-server/config.xml'.
clickhouse-01         | Merging configuration file '/etc/clickhouse-server/config.d/config.xml'.
clickhouse-01         | Logging debug to /var/log/clickhouse-server/clickhouse-server.log
clickhouse-01         | Logging errors to /var/log/clickhouse-server/clickhouse-server.err.log
clickhouse-02         | Processing configuration file '/etc/clickhouse-server/config.xml'.
clickhouse-02         | Merging configuration file '/etc/clickhouse-server/config.d/config.xml'.
clickhouse-02         | Saved preprocessed configuration to '/var/lib/clickhouse/preprocessed_configs/config.xml'.
clickhouse-02         | Processing configuration file '/etc/clickhouse-server/users.xml'.
clickhouse-02         | Merging configuration file '/etc/clickhouse-server/users.d/users.xml'.
clickhouse-02         | Saved preprocessed configuration to '/var/lib/clickhouse/preprocessed_configs/users.xml'.
clickhouse-01         | Processing configuration file '/etc/clickhouse-server/config.xml'.
clickhouse-01         | Merging configuration file '/etc/clickhouse-server/config.d/config.xml'.
clickhouse-01         | Saved preprocessed configuration to '/var/lib/clickhouse/preprocessed_configs/config.xml'.
clickhouse-01         | Processing configuration file '/etc/clickhouse-server/users.xml'.
clickhouse-01         | Merging configuration file '/etc/clickhouse-server/users.d/users.xml'.
clickhouse-01         | Saved preprocessed configuration to '/var/lib/clickhouse/preprocessed_configs/users.xml'.
```

nellicus commented:
ha! good catch, anyway was same setup with my logging changes and chproxy. keeper logs to logs vs stdou even better, make a commit? let-s quick sync on slack so we close this

DanRoscigno commented on docker-compose-recipes/cluster_1S_2R/fs/volumes/clickhouse-keeper-02/etc/clickhouse-keeper/keeper_config.xml:
```diff
@@ -0,0 +1,31 @@
+<clickhouse>
```
```suggestion
<clickhouse>
    <logger>
        <level>trace</level>
        <log>/var/log/clickhouse-keeper/clickhouse-keeper.log</log>
        <errorlog>/var/log/clickhouse-keeper/clickhouse-keeper.err.log</errorlog>
        <size>1000M</size>
        <count>3</count>
    </logger>

```

DanRoscigno commented on docker-compose-recipes/cluster_1S_2R/fs/volumes/clickhouse-keeper-03/etc/clickhouse-keeper/keeper_config.xml:
```diff
@@ -0,0 +1,31 @@
+<clickhouse>
```
```suggestion
<clickhouse>
    <logger>
        <level>trace</level>
        <log>/var/log/clickhouse-keeper/clickhouse-keeper.log</log>
        <errorlog>/var/log/clickhouse-keeper/clickhouse-keeper.err.log</errorlog>
        <size>1000M</size>
        <count>3</count>
    </logger>

```

nellicus commented:
@DanRoscigno good to push keeper logs to file, but trace level by default is overkill, so I'd switch to  `information`

adding to that, for keepers right now we have discrepancies between:

volume:

```
volumes:
      - ${PWD}/fs/volumes/clickhouse-keeper-03/etc/clickhouse-keeper/keeper_config.xml:/etc/clickhouse-keeper/keeper_config.xml
      - ${PWD}/fs/volumes/clickhouse-keeper-03/var/lib/clickhouse:/var/lib/clickhouse
      - ${PWD}/fs/volumes/clickhouse-keeper-03/var/log/clickhouse-server:/var/log/clickhouse-server
```

xml config

```
    <logger>
        <level>information</level>
        <log>/var/log/clickhouse-keeper/clickhouse-keeper.log</log>
        <errorlog>/var/log/clickhouse-keeper/clickhouse-keeper.err.log</errorlog>
        <size>1000M</size>
        <count>3</count>
    </logger>
    <listen_host>0.0.0.0</listen_host>
    <keeper_server>
        <tcp_port>9181</tcp_port>
        <server_id>1</server_id>
        <log_storage_path>/var/lib/clickhouse/coordination/log</log_storage_path>
        <snapshot_storage_path>/var/lib/clickhouse/coordination/snapshots</snapshot_storage_path>
```


and folder structure

```
➜  cluster_1S_2R git:(docker-compose-import) ✗ pwd
/Users/abonuccelli/Projects/ClickHouse/examples/docker-compose-recipes/cluster_1S_2R
➜  cluster_1S_2R git:(docker-compose-import) ✗ find fs/volumes/clickhouse-keeper-01/var/log
fs/volumes/clickhouse-keeper-01/var/log
fs/volumes/clickhouse-keeper-01/var/log/clickhouse-server
fs/volumes/clickhouse-keeper-01/var/log/.gitignore
➜  cluster_1S_2R git:(docker-compose-import) ✗ 
```

will try to fix these by today

nellicus commented:
@DanRoscigno  I've pushed [multiple](https://github.com/ClickHouse/examples/pull/2/commits/96da9f604395fdfd67ca878788111b5835a2e46d) fixes:
- add proper display names for grafana, openldap, S3 minio;
- fixed paths for keeper logs
- set keeper logging level to `information`

nellicus commented:
I also think that we should be mapping just the strictly necessary volumes to feed config files, and otherwise leave everything in the container fs, user can then be free to map whaterver is needed. this also dramatically reduces the complexity and maintenance burden of these recipes (I have plenty more coming). So I'll work on that next

nellicus commented:
I've also added all cluster_* variants with CH proxy and reorganised the repo and README.md(s)

DanRoscigno commented:
It looks like the keeper_config.xml got replaced with the generated one.  Does this improve things?  I worry about giving people a file that has the "Do not edit this file" header, as ideally they should experiment to learn.

nellicus commented:
@DanRoscigno I am not sure I follow, can you please be more specific?
which of the 6 recipes using keepers are you referring to?
looking at `cluster_2S_1R`, `clickhouse-keeper-01` for example we do mount the [folder](https://github.com/ClickHouse/examples/pull/2/files#diff-091dff850187dd996ea093fcd0fa28e3e1d92a308680f75ae734076e2951ad4dR48) to make use of the related [keeper_config.xml](https://github.com/ClickHouse/examples/pull/2/files#diff-c71a6a37e0ecb24c600bf0ffa0d213577a021a5734425a600f5e596dfed166b4) as have discussed in previous iterations, and use same approach throughout all occurrences.

I did test each of the 9 recipes provided by launching them and creating a replicated table (for the `cluster_*` ones) FWIW

DanRoscigno commented:
sorry, cluster_1S_2R and 1S_2R with proxy

nellicus commented:
I've just changed to mount for all CH components just the config files, not the folders, preprocessed files will hence not make it back into host fs.
also fixed some leftover values still referring to previous subnetting

nellicus commented:
@DanRoscigno I didn't realise earlier that preprocessed file made it into the repo

nellicus commented:
attempting to get rid of all network references and script, and hardcoded IPs and use hostnames to simplify cc @DerekChia
